### PR TITLE
[RLlib] Fix priority update for sequenced batches

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Signed-off-by: Artur Niederfahrenhorst <artur@anyscale.com>

## Why are these changes needed?

I discovered that RNNSAC is broken because it requires updates of the priorities of sequenced batches.
This has gone undiscovered because the compilation and the example test both never start learning, because we don't set the learning_starts parameter appropriately. Learning is never checked. Changing our tests to catch errors on complete training steps is part of another PR.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/replay_buffers/utils.py [hunk 1]
```diff
@@ -52,11 +53,34 @@ def update_priorities_in_replay_buffer(
...
+                # priorities, we need one index per td_error
+                _batch_indices = []
+
+                # Sequenced batches have been zero padded to max_seq_len.
+                # Depending on how batches are split during learning, not all
+                # sequences have an associated td_error (trailing ones missing).
+                if policy_batch.zero_padded:
+                    seq_lens = len(td_error) * [policy_batch.max_seq_len]
+                else:
+                    seq_lens = policy_batch[SampleBatch.SEQ_LENS][: len(td_error)]
```
Comment (sven1977) [inline rllib/utils/replay_buffers/utils.py:75]: Just curious:
What would lead to a situation where:
```
policy_batch[SampleBatch.SEQ_LENS][: len(td_error)] != policy_batch[SampleBatch.SEQ_LENS]
```

**Comment by sven1977** [comment]
Hey @ArturNiederfahrenhorst , thanks for the PR and great debugging!

Just one question on the logic. Otherwise, looks good.

File: rllib/utils/replay_buffers/utils.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/utils.py:75]: I was wondering thinking about this a lot because `len(td_error)` never seemed to fit the sequence lengths. Even if I took into account `max_seq_len` because of padding.

The problem in RNNSAC is: It uses SAC's train_batch_size of 256. While padding the sequence to prepare it for feeding it into an RNN, we increase the size (to 264 in this example). This blows up our sequences and we end up having an incomplete last sequence to train on because, before learning, we cut the batch to 256. Therefore the last sequence would always be incomplete.

Screenshot 2022-08-09 at 18 04 22

In our Policies, we cut loaded batches if they are too long. For example this snippet is from our current TorchPolicyV2:
 ```
             if device_batch_size >= len(self._loaded_batches[0][0]):
                batch = self._loaded_batches[0][0]
            else:
                batch = self._loaded_batches[0][0][offset : offset + device_batch_size]
```

So if you were to learn on a batch that has sequences, but is not padded (for whatever reason), and is larger than the device_batch_size, the produced td_error would be smaller than the number of sequences. Therefore, the discarded sequences would not produce a TD error and so `len(td_error)!=len(batch.SEQ_LENS)`.

**Review by sven1977 — APPROVED** [review]
Thanks again @ArturNiederfahrenhorst . Looks good to me.
